### DISCUSSION TITLE
Passing source file compiler arguments to a custom_target

### DISCUSSION THREAD

**Discussion by akbyrd** [mesonbuild/meson#12522] [Q&A]
I'm learning Meson and practicing some scenarios I expect to need to support. One such scenario is running a code generation step on C++ source files. I want to run a compiler pre-pass over source files, parse some information, and output a new C++ file to compile in addition to the original. I need to know the set of compiler arguments that will be used with the original file when I perform the pre-pass.

Input
```
src/
  - foo.cpp
```

Output
```
builddir/
  - foo.generated.cpp
```

Here's what I have so far in my build file:
```
project(
	'engine',
	'cpp',
	version: '0.0.1',
	default_options: [
		'warning_level=3',
		'cpp_std=c++14'
	]
)

source_files = files(
	'foo.cpp'
)

generator_path    = '..' / 'build-system' / 'code-gen.ps1'
generated_targets = []

foreach source_file : source_files
	source_args = ...
	generated_target = custom_target(
		input: source_file,
		command: [ 'pwsh', '-File', generator_path, '@OUTPUT@', source_args ],
		output: [ '@BASENAME@.generated.cpp' ]
	)
	generated_targets += generated_target
endforeach

executable(
	'engine',
	sources: source_files + generated_targets
)
```

There's `get_option('cpp_args')`, but that's just extra arguments and not _all_ arguments.

In theory I could read `compile_commands.json` manually in the code-gen script. However, this means every `custom_target` invocation has to re-parse a json file.

**Comment by mitchgrout** [comment]
I don't believe there is a valid way to get at the raw compile arguments at this time. Could you provide some more information on what the pre-processing step you need to do is? We may be able to find an alternative way to approach this within the existing system

**Reply by akbyrd** [reply]
Sure. The main use case is generating serialization and introspection code for C++ structs. I parse the file, find types that opt-in, and generate a new file with the desired code. Being C++ it's necessary to know essentially the full set of compiler args to be able to handle preprocessor definitions, include paths, etc.

I have full control of the system, so I can change it to play nicely with Meson if there is a suitable approach available.
